python_modules/dagster/dagster/_config/structured_config.py

```diff
@@ -219,6 +258,36 @@ def create_io_manager_to_pass_to_user_code(self, context) -> IOManager:
         return self
 
 
+def is_structured_io_manager_cls(resource: Any) -> bool:
+    try:
+        return issubclass(resource, StructuredConfigIOManagerBase)
+    except TypeError:
+        pass
+    return False
+
+
+class UnconfiguredStructuredIOManager(IOManagerDefinition):
+    def __init__(
+        self,
+        io_manager: Type[StructuredConfigIOManagerBase],
+    ):
+        schema = infer_schema_from_config_class(io_manager)
+
+        def resource_fn(context: InitResourceContext):
+            instantiated = io_manager(**context.resource_config)
+
+            if is_io_manager_context_provided(instantiated):
+                return instantiated.resource_fn(context)
+            else:
+                return instantiated.resource_fn()
```

Comment:
Seems like we can eliminate a bunch of code duplication here by making `is_io_manager_context_provided` and `is_context_provided`. Might have to do some shenanigans for make pyright happy but this feels like too much duplication between `UnconfiguredStructuredIOManager` and `UnconfiguredStructuredResource` to me. Even just a helpful function.

Reply:
tidied up the implementations a bit, which lets us remove the context guards